src/hotspot/os/linux/os_linux.cpp

```diff
@@ -4821,6 +4824,25 @@ jint os::init_2(void) {
     }
   }
 
+  // Check the availability of MADV_POPULATE_WRITE.
+  if (UseMadvPopulateWrite) {
+    // Some downstream kernels recognize MADV_POPULATE_WRITE_value as another
+    // advice, so the check of versions is required here.
+    // See https://github.com/oracle/linux-uek/issues/23
+    long major, minor;
+    Linux::kernel_version(&major, &minor);
+    bool supportMadvPopulateWrite =
+      ((major > 5 || (major == 5 && minor >= 14)) &&
+       (::madvise(0, 0, MADV_POPULATE_WRITE) == 0));
```

Comment:
Shouldn't this be checking ` MADV_POPULATE_WRITE_value `?

I still can't get my head around the possible compile time values versus the potential runtime interpretation of the ` MADV_POPULATE_WRITE`  value.

Reply:
For example, Linux 5.4-UEK does not support populating pages, and `MADV_POPULATE_WRITE` is not defined there. However, the kernel internally interprets 23 as `MADV_DONEXEC`, while 23 is a value used by upstream kernels as `MADV_POPULATE_WRITE`. `MADV_POPULATE_WRITE_value` is used by hotspot internally, and used for compiling JDK properly when `MADV_POPULATE_WRITE` is not defined. `MADV_POPULATE_WRITE_value` was suggested there: https://github.com/openjdk/jdk/pull/15781#discussion_r1345784162

Reply:
Additionally, If `MADV_POPULATE_WRITE` is defined but not the same as `MADV_POPULATE_WRITE_value`, there will be a build failure, so hotspot is guaranteed to use 23 here regardless of compile time or runtime. 23 may not be interpreted as `MADV_POPULATE_WRITE` by a downstream kernel. A known case is a downstream of pre-5.14 kernel, so the version of kernel is checked here. This patch does not have tricks on advice numbers.

Reply:
Okay the comment:
```
 // Some downstream kernels recognize MADV_POPULATE_WRITE_value as another
```
made me think we needed to check the "value" version but I see now they are already guaranteed to be the same.